src/lua/datetime.lua

```diff
@@ -0,0 +1,622 @@
+        -- if there is one of nsec, usec, msec provided
+        -- then ignore fraction in timestamp
+        -- otherwise - use nsec, usec, or msec
+        if count_usec == 0 then
+            nsec = fraction * 1e9
+        else
+            error('only integer values allowed in timestamp '..
+                  'if nsec, usec, or msecs provided', 2)
+        end
+
+        self.epoch = sec_int
+        self.nsec = nsec
+
+        return self
+    end
+
+    local offset0 = self.tzoffset
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset)
+        check_range(offset, -720, 720, 'tzoffset')
+    end
+
+    if obj.tz ~= nil then
+        nyi('tz')
+    end
+
+    -- .year, .month, .day
+    if ymd then
+        datetime_ymd_update(self, y or y0, M or M0, d or d0, offset or offset0)
+    end
+
+    -- .hour, .minute, .second
+    if hms then
+        datetime_hms_update(self, h or h0, m or m0, sec or sec0, offset or offset0)
+    end
+
+    self.epoch, self.nsec = normalize_nsec(self.epoch, self.nsec)
+
+    if offset ~= nil then
+        self.tzoffset = offset
+    end
+
+    return self
+end
+
+local function datetime_strftime(self, fmt)
+    local strfmt_sz = 128
+    local buff = ffi.new('char[?]', strfmt_sz)
+    check_str(fmt, "datetime.strftime()")
+    builtin.tnt_datetime_strftime(self, buff, strfmt_sz, fmt)
+    return ffi.string(buff)
+end
+
+local function datetime_format(self, fmt)
+    if fmt ~= nil then
+        return datetime_strftime(self, fmt)
+    else
+        return datetime_tostring(self)
+    end
+end
+
+local datetime_index_fields = {
+        timestamp = function(self) return self.epoch + self.nsec / 1e9 end,
+
+        year = function(self) return builtin.tnt_dt_year(local_dt(self)) end,
+        yday = function(self) return builtin.tnt_dt_doy(local_dt(self)) end,
+        month = function(self) return builtin.tnt_dt_month(local_dt(self)) end,
+        day = function(self)
+            return builtin.tnt_dt_dom(local_dt(self))
+        end,
+        wday = function(self)
+            return dow_to_wday(builtin.tnt_dt_dow(local_dt(self)))
+        end,
+        hour = function(self) return math_floor((local_secs(self) / 3600) % 24) end,
+        min = function(self) return math_floor((local_secs(self) / 60) % 60) end,
+        sec = function(self) return self.epoch % 60 end,
+        usec = function(self) return self.nsec / 1e3 end,
+        msec = function(self) return self.nsec / 1e6 end,
+}
+
+local datetime_index_functions = {
+        format = datetime_format,
+        totable = datetime_totable,
+        set = datetime_set,
+}
+
+local function datetime_index(self, key)
+    local handler_field = datetime_index_fields[key]
+    return handler_field ~= nil and handler_field(self) or
+           datetime_index_functions[key]
+end
+
+ffi.metatype(datetime_t, {
+    __tostring = datetime_tostring,
+    __serialize = datetime_serialize,
```

Comment:
Realized that `__serialize` does not work directly from metatable, but only works via `__index` table. So here is a patch fixing `__serialize` and also fixing extra indents we used to have there around:

```diff
diff --git a/src/lua/datetime.lua b/src/lua/datetime.lua
index c19e4e742..6ab7d5ae8 100644
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -570,28 +570,29 @@ local function datetime_format(self, fmt)
 end
 
 local datetime_index_fields = {
-        timestamp = function(self) return self.epoch + self.nsec / 1e9 end,
-
-        year = function(self) return builtin.tnt_dt_year(local_dt(self)) end,
-        yday = function(self) return builtin.tnt_dt_doy(local_dt(self)) end,
-        month = function(self) return builtin.tnt_dt_month(local_dt(self)) end,
-        day = function(self)
-            return builtin.tnt_dt_dom(local_dt(self))
-        end,
-        wday = function(self)
-            return dow_to_wday(builtin.tnt_dt_dow(local_dt(self)))
-        end,
-        hour = function(self) return math_floor((local_secs(self) / 3600) % 24) end,
-        min = function(self) return math_floor((local_secs(self) / 60) % 60) end,
-        sec = function(self) return self.epoch % 60 end,
-        usec = function(self) return self.nsec / 1e3 end,
-        msec = function(self) return self.nsec / 1e6 end,
+    timestamp = function(self) return self.epoch + self.nsec / 1e9 end,
+
+    year = function(self) return builtin.tnt_dt_year(local_dt(self)) end,
+    yday = function(self) return builtin.tnt_dt_doy(local_dt(self)) end,
+    month = function(self) return builtin.tnt_dt_month(local_dt(self)) end,
+    day = function(self)
+        return builtin.tnt_dt_dom(local_dt(self))
+    end,
+    wday = function(self)
+        return dow_to_wday(builtin.tnt_dt_dow(local_dt(self)))
+    end,
+    hour = function(self) return math_floor((local_secs(self) / 3600) % 24) end,
+    min = function(self) return math_floor((local_secs(self) / 60) % 60) end,
+    sec = function(self) return self.epoch % 60 end,
+    usec = function(self) return self.nsec / 1e3 end,
+    msec = function(self) return self.nsec / 1e6 end,
 }
 
 local datetime_index_functions = {
-        format = datetime_format,
-        totable = datetime_totable,
-        set = datetime_set,
+    format = datetime_format,
+    totable = datetime_totable,
+    set = datetime_set,
+    __serialize = datetime_serialize,
 }
 
 local function datetime_index(self, key)
@@ -602,7 +603,6 @@ end
 
 ffi.metatype(datetime_t, {
     __tostring = datetime_tostring,
-    __serialize = datetime_serialize,
     __eq = datetime_eq,
     __lt = datetime_lt,
     __le = datetime_le,
```